graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,358 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.graphql;
+
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.PublisherBasedStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.websocket.WebSocketWriter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import graphql.ErrorClassification;
+import graphql.ExecutionInput;
+import graphql.ExecutionResult;
+import graphql.GraphQLError;
+import graphql.language.SourceLocation;
+
+/**
+ * Handles the graphql-ws sub protocol within a web socket.
+ * Handles potentially multiple subscriptions through the same web socket.
+ */
+class GraphqlWSSubProtocol {
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlWSSubProtocol.class);
+    private final GraphqlExecutor graphqlExecutor;
+    private final HashMap<String, ExecutionResultSubscriber> graphqlSubscriptions = new HashMap<>();
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final TypeReference<Map<String, Object>> JSON_MAP =
+            new TypeReference<Map<String, Object>>() {};
+
+    private boolean connectionInitiated;
+
+    private final ServiceRequestContext ctx;
+    private final Map<String, Object> upgradeCtx;
+    private Map<String, Object> connectionCtx = ImmutableMap.of();
+
+    GraphqlWSSubProtocol(ServiceRequestContext ctx, GraphqlExecutor executor) {
+        upgradeCtx = GraphqlServiceContexts.graphqlContext(ctx);
+        graphqlExecutor = executor;
+        this.ctx = ctx;
+    }
+
+    /**
+     * Receives an event and returns a response if one should be sent.
+     */
+    @Nullable
+    public void handle(String event, WebSocketWriter out) {
+        try {
+            final Map<String, Object> eventMap = parseJsonString(event, JSON_MAP);
+            final String type = toStringFromJson("type", eventMap.get("type"));
+            requireNonNull(type, "type");
```

Comment:
How about sending the close with 4400 status when an exception is raised while parsing the event?
https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#invalid-message
I think we also have to call `subscription.cancel` when it fails to process.

In order to do that, how about just merging `GraphqlWebSocketSubscriber` and `GraphqlWSSubProtocol` into one?
`GraphqlWebSocketSubscriber` doesn't have much logic in it so I think we can merge those.

Reply:
Personally, I'd prefer not. `graphql-ws` is not the first protocol, and I doubt it will be the last.

About the 4400 status - we are closing the websocket on such a failure. So it's not obvious to me that the client would ever have the chance to handle the GraphQL error before the websocket closes on them.